src/Stack/Nix.hs

```diff
@@ -0,0 +1,146 @@
+import           Data.Typeable
+import           Data.Version (showVersion)
+import           Network.HTTP.Client.Conduit (HasHttpManager)
+import qualified Paths_stack as Meta
+import           Prelude -- Fix redundant import warnings
+import           Stack.Constants (stackProgName)
+import           Stack.Docker (reExecArgName)
+import           Stack.Exec (exec)
+import           Stack.Types
+import           Stack.Types.Internal
+import           System.Environment (lookupEnv,getArgs,getExecutablePath)
+import           System.Exit (exitSuccess, exitWith)
+
+
+-- | If Nix is enabled, re-runs the currently running OS command in a Nix container.
+-- Otherwise, runs the inner action.
+reexecWithOptionalShell
+    :: M env m
+    => IO ()
+    -> m ()
+reexecWithOptionalShell inner =
+  do config <- asks getConfig
+     inShell <- getInShell
+     isReExec <- asks getReExec
+     if nixEnable (configNix config) && not inShell && not isReExec
+       then runShellAndExit getCmdArgs
+       else liftIO (inner >> exitSuccess)
+  where
+    getCmdArgs = do
+        args <-
+            fmap
+                (("--" ++ reExecArgName ++ "=" ++ showVersion Meta.version) :)
+                (liftIO getArgs)
+        exePath <- liftIO getExecutablePath
+        return (exePath, args)
+
+runShellAndExit :: M env m
+                => m (String, [String])
+                -> m ()
+runShellAndExit getCmdArgs = do
+     config <- asks getConfig
+     (cmnd,args) <- getCmdArgs
+     let mshellFile = nixInitFile (configNix config)
+         pkgsInConfig = nixPackages (configNix config)
+     if not (null pkgsInConfig) && isJust mshellFile then
+       throwM NixCannotUseShellFileAndPackagesException
+       else return ()
+     let nixopts = case mshellFile of
+           Just filePath -> [filePath]
+           Nothing -> ["-E", T.unpack $ T.intercalate " " $ concat
+                              [["with (import <nixpkgs> {});"
+                               ,"runCommand \"myEnv\" {"
+                               ,"buildInputs=lib.optional stdenv.isLinux glibcLocales ++ ["],pkgsInConfig,["];"
+                               ,T.pack inShellEnvVar,"=1 ;"
+                               ,"STACK_IN_NIX_EXTRA_ARGS=''"]
+                               ,      (map (\p -> T.concat
+                                                  ["--extra-lib-dirs=${",p,"}/lib"
+                                                  ," --extra-include-dirs=${",p,"}/include "])
+                                           pkgsInConfig), ["'' ;"
+                               ,"} \"\""]]]
+                    -- glibcLocales is necessary on Linux to avoid warnings about GHC being incapable to set the locale.
+         fullArgs = concat [ -- ["--pure"],
+                            map T.unpack (nixShellOptions (configNix config))
+                           ,nixopts
+                           ,["--command", intercalate " " (map (\a -> "'"++a++"'") (cmnd:args))
```

Comment:
Surrounding by `'` isn't quite enough either; what if the value contain a `'` itself?  A reliable way to quote all characters, at least for bourne shell derivatives, is to surround the values in `'`s (like you did) and also replace all occurrences of `'` within each value with `'"'"'`.  Example code in Propellor's [shellEscape](http://hackage.haskell.org/package/propellor-2.14.0/docs/Propellor-Property-Cmd.html#v:shellEscape).

Of course it would be much nicer if `nix-shell` gave you a way to do this without having to escape it at all, but I guess that's not possible.